src/main/java/org/haffson/adventofcode/days/day07/StepSorter.java

```diff
@@ -0,0 +1,56 @@
+package org.haffson.adventofcode.days.day07;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class StepSorter {
+    @Nonnull
+    private final SortedSteps sortedSteps;
+
+    public StepSorter() {
+        this.sortedSteps = new SortedSteps();
+    }
+
+    @Nonnull
+    public String getSortedSteps(@Nonnull final List<Step> stepsInput) {
+        requireNonNull(stepsInput, "stepsInput");
+
+        final List<Step> stepsCopy = new ArrayList<>(stepsInput);
+
+        stepsInput.forEach(step -> addNextStepToResult(stepsCopy));
+
+        return sortedSteps.getStepsAsString();
+    }
+
+
+    private void addNextStepToResult(@Nonnull final List<Step> steps) {
+        requireNonNull(steps, "steps");
+
+        final List<Step> availableSteps = getAvailableSteps(steps);
+        final List<Step> sortedAvailableSteps = new ArrayList<>(availableSteps);
+        sortedAvailableSteps.sort(Comparator.comparing(Step::stepName));
+        final Step nextStep = sortedAvailableSteps.get(0);
+        sortedSteps.addStep(nextStep);
+        steps.remove(nextStep);
+    }
+
+    @Nonnull
+    private List<Step> getAvailableSteps(@Nonnull final List<Step> steps) {
+        requireNonNull(steps, "steps");
+
+        final List<Step> availableSteps = new ArrayList<>();
+
+        steps.forEach(step -> {
+            if (new HashSet<>(sortedSteps.getCharacterSequence()).containsAll(step.dependsOn())) {
```

Comment:
Warum wird hier in ein `HashSet` konvertiert?

Reply:
IntelliJ hatte mir folgende Warnung gegeben: 

> Call to 'list.containsAll(collection)' may have poor performance

Der HashSet-Wrapper soll dies wohl verbessern

Reply:
Interessant, hab ich bei mir noch nie gesehen.

Allein die Kopie der Liste in ein Set dürfte bei O(n) liegen. Es würde mich etwas wundern, wenn sich das nach einem `containsAll()`-Aufruf schon wieder auszahlt